converter/src/main/java/gov/cms/qpp/conversion/decode/QrdaDecoderEngine.java

```diff
@@ -0,0 +1,238 @@
+
+import gov.cms.qpp.conversion.Context;
+import gov.cms.qpp.conversion.correlation.PathCorrelator;
+import gov.cms.qpp.conversion.model.DecodeData;
+import gov.cms.qpp.conversion.model.Decoder;
+import gov.cms.qpp.conversion.model.Node;
+import gov.cms.qpp.conversion.model.Registry;
+import gov.cms.qpp.conversion.model.TemplateId;
+import gov.cms.qpp.conversion.segmentation.QrdaScope;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Top level Decoder for parsing into QPP format.
+ */
+public class QrdaDecoderEngine extends XmlDecoderEngine {
+
+	private static final Logger DEV_LOG = LoggerFactory.getLogger(QrdaDecoderEngine.class);
+	private static final String TEMPLATE_ID = "templateId";
+	private static final String NOT_VALID_QRDA_III_FORMAT = "The file is not a QRDA-III XML document";
+	private static final String ROOT_STRING = "root";
+	private static final String EXTENSION_STRING = "extension";
+
+	protected final Context context;
+	private final Set<TemplateId> scope;
+	private final Registry<QrdaDecoder> decoders;
+
+	/**
+	 * Initialize a qpp xml decoder
+	 */
+	public QrdaDecoderEngine(Context context) {
+		Objects.requireNonNull(context, "converter");
+
+		this.context = context;
+		this.scope = context.hasScope() ? QrdaScope.getTemplates(context.getScope()) : null;
+		this.decoders = context.getRegistry(Decoder.class);
+	}
+
+	/**
+	 * Decodes the top of the XML document
+	 *
+	 * @param xmlDoc XML Document to be parsed
+	 * @return Root node
+	 */
+	@Override
+	public Node decode(Element xmlDoc) {
+		Node rootNode = new Node();
+		Element rootElement = xmlDoc.getDocument().getRootElement();
+		defaultNs = rootElement.getNamespace();
+
+		rootNode.setType(TemplateId.PLACEHOLDER);
+		rootNode.setPath(XPathHelper.getAbsolutePath(rootElement));
+
+		QrdaDecoder rootDecoder = null;
+		for (Element element : rootElement.getChildren(TEMPLATE_ID, rootElement.getNamespace())) {
+			rootDecoder = getDecoder(getTemplateId(element));
+			if (rootDecoder != null) {
+				break;
+			}
+		}
+
+		if (rootDecoder != null) {
+			rootNode = this.decodeTree(rootElement, rootNode).getNode().getChildNodes().get(0);
+		} else {
+			rootNode = this.decodeTree(rootElement, rootNode).getNode();
+		}
+
+		return rootNode;
+	}
+
+	private DecodeData decodeTree(final Element element, final Node parentNode) {
+		DecodeData result = decodeSingleElement(element, parentNode);
+		DecodeResult decodedResult = result.getDecodeResult();
+		Node decodedNode = result.getNode();
+
+		decodedNode = decideNewParentNode(decodedNode, parentNode);
+
+		if (DecodeResult.TREE_FINISHED == decodedResult) {
+			return new DecodeData(DecodeResult.TREE_FINISHED, decodedNode);
+		} else if (DecodeResult.TREE_ESCAPED == decodedResult) {
+			return new DecodeData(DecodeResult.TREE_FINISHED, null);
+		}
+
+		return decodeChildren(element, decodedNode);
+	}
+
+	private DecodeData decodeSingleElement(final Element element, final Node parentNode) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 4, down to the maximum allowed 3. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1142)